markdown/ch3.markdown

```diff
@@ -528,26 +574,58 @@ Above, we saw how we can combine the output of two different filters, the
 horizontal Sobel filter, and the vertical one. Can you write a function that
 does this in a single pass using `ndi.generic_filter`?
 
+<!-- solution begin -->
+
+```python
+hsobel = np.array([[ 1,  2,  1],
+                   [ 0,  0,  0],
+                   [-1, -2, -1]])
+
+vsobel = hsobel.T
+
+hsobel_r = np.ravel(hsobel)
+vsobel_r = np.ravel(vsobel)
+
+def sobel_magnitude_filter(values):
+    h_edge = values @ hsobel_r
+    v_edge = values @ vsobel_r
+    return np.hypot(h_edge, v_edge)
+```
+
+Now we can try it out on the coins image:
+
+```python
+sobel_mag = ndi.generic_filter(coins, sobel_magnitude_filter, size=3)
+plt.imshow(sobel_mag)
+```
+
+<!-- solution end -->
+
 <!-- exercise end -->
 
 ## Graphs and the NetworkX library
 
+Graphs are a natural representation for a bewildering variety of data. Pages on
+the world wide web, for example, can comprise nodes, while links between those
+pages can be, well, links. Or, in so-called *transcription networks*, nodes
+represent genes and edges connect genes that have a direct influence on each
+other's expression.
+
+> **Note: graphs and networks {.callout}**
+>
+> In this context, the term "graph" is synonymous with "network", not with
+> "plot".  Mathematicians and computer scientists invented slightly different
+> words to discuss these: graph = network, vertex = node, and edge = link.  As
+> most people do, we will be using these terms interchangeably.
+>
+> You might be slightly more familiar with the network terminology: a network
+> consists of *nodes* and *links* between the nodes. Equivalently, a graph
+> consists of *vertices* and *edges* between the vertices. In NetworkX, you
+> have `Graph` objects consisting of `nodes` and `edges` between the nodes. Oh
+> well!
```

Comment:
In my experience, nodes and vertices are commonly used terms, whereas I have never seen link used instead of edge (Wikipedia mentions arc / line but not link).

So, I think nodes + edges is perfectly consistent with common usage.